packages/orm/src/reactiveHints.ts

```diff
@@ -318,44 +318,42 @@ function maybeAddTypeFilterSuffix(
-      current.flatMap((c: any) => {
-        const currentValuePromise = maybeApplyTypeFilter(c[fieldName].load(), viaType);
-        // If we're going from Book.author back to Author to re-validate the Author.books collection,
-        // see if Book.author has changed, so we can re-validate both the old author's books and the
-        // new author's books.
-        const fieldKind = getMetadata(c).fields[fieldName]?.kind;
-        const isReference = fieldKind === "m2o" || fieldKind === "poly";
-        const isManyToMany = fieldKind === "m2m";
-        const changed = isChangeableField(c, fieldName) ? (c.changes[fieldName] as FieldStatus<any>) : undefined;
-        // See jsdoc comment about why this is only necessary for references...
-        if (isReference && changed && changed.hasUpdated && changed.originalValue) {
-          return [
-            currentValuePromise,
-            maybeApplyTypeFilter((changed as ManyToOneFieldStatus<any>).originalEntity, viaType),
-          ];
-        }
-        if (isManyToMany) {
-          const m2m = c[fieldName] as ManyToManyCollection<any, any>;
-          const joinRows = getEmInternalApi(m2m.entity.em).joinRows(m2m);
-          // Return a tuple of [currentRows, removedRows]
-          return [currentValuePromise, joinRows.removedFor(m2m, c)];
-        }
-        return [currentValuePromise];
-      }),
-    );
-    // Use flat() to get them all as entities
-    current = entitiesOrLists.flat().filter((e) => e !== undefined);
+    for (const c of current as Set<any>) {
+      const currentValuePromise = maybeApplyTypeFilter(c[fieldName].load(), viaType);
+      // Always wait for the relation itself
+      promises.push(currentValuePromise);
+      // If we're going from Book.author back to Author to re-validate the Author.books collection,
+      // see if Book.author has changed, so we can re-validate both the old author's books and the
+      // new author's books.
+      const fieldKind = getMetadata(c).fields[fieldName]?.kind;
+      const isReference = fieldKind === "m2o" || fieldKind === "poly";
+      const isManyToMany = fieldKind === "m2m";
+      const changed = isChangeableField(c, fieldName) ? (c.changes[fieldName] as FieldStatus<any>) : undefined;
+      // See jsdoc comment about why this is only necessary for references...
+      if (isReference && changed && changed.hasUpdated && changed.originalValue) {
+        promises.push(maybeApplyTypeFilter((changed as ManyToOneFieldStatus<any>).originalEntity, viaType));
+      }
+      if (isManyToMany) {
+        const m2m = c[fieldName] as ManyToManyCollection<any, any>;
+        const joinRows = getEmInternalApi(m2m.entity.em).joinRows(m2m);
+        // Return a tuple of [currentRows, removedRows]
+        promises.push(joinRows.removedFor(m2m, c));
+      }
+    }
+    const nextLevel = await Promise.all(promises);
+    // Make a new set so that we dedupe as we go
+    current = flatAndUnique(nextLevel);
```

Comment:
One of the actual changes here is that we `new Set`/dedupe entities as we crawl through the hint, instead of just at the end, i.e. for `template.items.options.foo`, the items will be ~1000 some items, but but only ~40 unique options across them, so if we unique as we go, we only call `foo.load()` 40 times instead of ~1000.